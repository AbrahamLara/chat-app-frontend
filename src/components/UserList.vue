<template>
  <v-list class="user-list pa-0 d-flex flex-column fill-height">
    <v-list-item dense>
      <v-list-item-title v-text="name"></v-list-item-title>
    </v-list-item>
    <v-divider></v-divider>
    <div v-if="loading" class="d-flex flex align-center justify-center mt-5">
      <v-progress-circular
        size="50"
        color="primary"
        indeterminate
      ></v-progress-circular>
    </div>
    <template v-else-if="error">
      <div
        class="user-list__error-message mt-3 d-flex justify-center flex-column"
      >
        <v-icon large color="error">mdi-alert-circle-outline</v-icon>
        <div class="text--secondary text-center" v-text="error"></div>
      </div>
    </template>
    <template v-else>
      <div class="user-list__items-container overflow-hidden flex">
        <div class="user-list__items-wrapper fill-height overflow-auto">
          <slot></slot>
        </div>
      </div>
    </template>
  </v-list>
</template>

<script lang="ts">
import Vue from 'vue';
import { Component, Prop, PropSync } from 'vue-property-decorator';

@Component({
  name: 'UserListItem',
})
export default class UserList extends Vue {
  @Prop({ required: true }) name!: string;
  @Prop({ default: false }) loading!: boolean;
  @PropSync('errorMessage', { type: String, default: '' })
  error!: string;
}
</script>

<style lang="sass">
.user-list
  .v-list-item:first-child
    flex: unset

  .user-list-item:last-of-type + .v-divider
    display: none
</style>
