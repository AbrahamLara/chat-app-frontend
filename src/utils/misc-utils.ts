/**
 * An user item that can be used for contexts such as displaying a user search or chat result.
 */
import Vue from 'vue';

export interface UserItem {
  /**
   * The name of the user.
   */
  name: string;

  /**
   * The timestamp to determine the duration of time relative to today.
   */
  timeStamp?: string | number;

  /**
   * A user message.
   */
  message?: string;
}

/**
 * Returns a cookie stored in the page with the given name.
 */
export function getBrowserCookie(name: string) {
  return Vue.$cookies.get(name);
}
