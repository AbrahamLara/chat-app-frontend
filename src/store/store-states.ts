import { THEME } from '@/utils/theme-utils';
import { FormAlertMessage, AlertMessage } from '@/utils/alerts-utils';
import { UserChatItem } from '@/utils/chat-utils';

/**
 * Describes the state of application.
 */
interface AppState {
  /**
   * The current theme to render the app in. Either dark or light.
   */
  theme: THEME;

  /**
   * The list of chat groups the user is in.
   */
  chats?: ChatsState;

  /**
   * The user's authenticated state.
   */
  auth?: UserState;
}

interface ChatsState extends AlertsState {
  /**
   * Describes the state of the user's chat list.
   */
  chats: UserChatItem[];
}

/**
 * Describes the user's authenticated state.
 */
interface UserState extends AlertsState {
  isAuthenticated: boolean;
}

/**
 * Describes the state of alert types on the page.
 */
interface AlertsState {
  alerts: {
    /**
     * Login form error messages.
     */
    errors: (AlertMessage | FormAlertMessage)[];

    /**
     * Register form error messages.
     */
    successes: AlertMessage[];
  };
}

export { AppState, ChatsState, AlertsState, UserState };
