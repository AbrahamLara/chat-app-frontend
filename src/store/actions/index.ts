import Vue from 'vue';
import { ActionTree } from 'vuex';
import { RootState } from '@/store/store-states';
import { THEME } from '@/utils/theme-utils';
import { loginUser, registerUser } from '@/api/auth-api';
import { namespaceAlerts } from '@/store/modules';
import { createAlertMessage, isFormErrorMessage } from '@/utils/alerts-utils';
import { LoginFormFields, RegisterFormFields } from '@/utils/auth-utils';
import {
  ADD_ERROR,
  ADD_SUCCESS,
  SET_ERRORS,
} from '@/store/constants/alerts-constants';
import {
  SET_THEME,
  LOGIN_USER,
  REGISTER_USER,
  SET_IS_AUTHENTICATED,
} from '@/store/constants/root-constants';

// Actions that affect how the application looks and behaves.
const appActions: ActionTree<RootState, RootState> = {
  [SET_THEME]: ({ commit }, theme: THEME) => {
    commit(SET_THEME, theme);
  },
  [LOGIN_USER]: async ({ commit }, loginForm: LoginFormFields) => {
    try {
      const res = await loginUser(loginForm);
      const data = await res.json();

      if (res.ok) {
        // Store the provided token as a cookie.
        // Set the cookie to expire at the same time as the cookie.
        Vue.$cookies.set('token', data.token, '3m');
        // Set the user as authenticated.
        commit(SET_IS_AUTHENTICATED, true);
      } else if (isFormErrorMessage(data)) {
        commit(namespaceAlerts(SET_ERRORS), data);
      } else {
        commit(namespaceAlerts(ADD_ERROR), data);
      }
    } catch (event) {
      const genericMessage = createAlertMessage(
        'An error occurred trying to sign you in!'
      );
      commit(namespaceAlerts(ADD_ERROR), genericMessage);
    }
  },
  [REGISTER_USER]: async ({ commit }, registerForm: RegisterFormFields) => {
    try {
      const res = await registerUser(registerForm);
      const data = await res.json();

      if (res.ok) {
        commit(namespaceAlerts(ADD_SUCCESS), data);
      } else if (isFormErrorMessage(data)) {
        commit(namespaceAlerts(SET_ERRORS), data);
      } else {
        commit(namespaceAlerts(ADD_ERROR), data);
      }
    } catch (event) {
      const genericMessage = createAlertMessage(
        'An error occurred trying to register you!'
      );
      commit(namespaceAlerts(ADD_ERROR), genericMessage);
    }
  },
  [SET_IS_AUTHENTICATED]: ({ commit }, isAuthenticated: boolean) => {
    commit(SET_IS_AUTHENTICATED, isAuthenticated);
  },
};

export default appActions;
