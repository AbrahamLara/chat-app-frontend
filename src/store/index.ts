import Vue from 'vue';
import Vuex from 'vuex';
import appActions from '@/store/actions';
import appMutations from '@/store/mutations';
import appModules from '@/store/modules';
import { THEME } from '@/utils/theme-utils';
import { AppState } from '@/store/store-states';

Vue.use(Vuex);

export const DEFAULT_ROOT_STATE: AppState = {
  theme: THEME.LIGHT,
};

export default new Vuex.Store<AppState>({
  state: { ...DEFAULT_ROOT_STATE },
  mutations: appMutations,
  actions: appActions,
  modules: appModules,
});
