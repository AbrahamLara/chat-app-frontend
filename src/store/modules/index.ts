import { ModuleTree } from 'vuex';
import { RootState } from '@/store/store-states';
import alertsModule from '@/store/modules/alerts-module';

export const ALERTS_NAMESPACE = 'alerts';

const appModules: ModuleTree<RootState> = {
  [ALERTS_NAMESPACE]: alertsModule,
};

/**
 * This function namespaces the provided mutation name and returns it.
 */
export function namespaceAlerts(mutation: string) {
  return `${ALERTS_NAMESPACE}/${mutation}`;
}

export default appModules;
