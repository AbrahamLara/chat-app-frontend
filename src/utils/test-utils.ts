import VueRouter, { RouteConfig } from 'vue-router';
import Vuex from 'vuex';
import { AppState } from '@/store/store-states';
import appMutations from '@/store/mutations';
import appActions from '@/store/actions';
import appModules from '@/store/modules';
import { DEFAULT_ROOT_STATE } from '@/store';
import { FormAlertMessage } from '@/utils/alerts-utils';
import { createLocalVue } from '@vue/test-utils';
import VueCookies from 'vue-cookies';
import Vue from 'vue';
import Vuetify from 'vuetify';

/**
 * Describes the available options for setting up a local vue instance for a test file.
 */
interface LocalVueSetupOptions {
  /**
   * Indicates that vuex plugin should be used.
   */
  useVuex?: boolean;

  /**
   * Indicates that vuetify plugin should be used.
   */
  useVuetify?: boolean;

  /**
   * Indicates that vue cookies plugin should be used.
   */
  useVueCookies?: boolean;

  /**
   * Indicates that vue router plugin should be used.
   */
  useVueRouter?: boolean;
}

/**
 * Resolves a vuetify warning of not being able to locate a data-app attribute on an element because vuetify
 * components should be wrapped by a v-app component.
 *
 * ```
 * Ex: [Vuetify] Unable to locate target [data-app] in "v-menu"
 * ```
 */
export function resolveVuetifyAppDataWarning() {
  const app = document.createElement('div');
  app.setAttribute('data-app', 'true');
  document.body.append(app);
}

/**
 * Create a Vuex store instance for testing.
 *
 * TODO: Reuse to create store in main.ts.
 *
 * @param vx The Vuex value
 * @param customState A state object that modifies the default state
 */
export function createStore(
  vx: typeof Vuex,
  customState: Partial<AppState> = {}
) {
  const mutations = appMutations;
  const actions = appActions;
  const modules = appModules;

  return new vx.Store({
    state: { ...DEFAULT_ROOT_STATE, ...customState },
    actions,
    mutations,
    modules,
  });
}

/**
 * Creates a VueRouter instance for testing.
 *
 * TODO: Reuse to create router in main.ts.
 *
 * @param vr The VueRouter value
 * @param routes The route config. If not provided, then a default config is provided.
 */
export function createRouter(
  vr: typeof VueRouter,
  routes: RouteConfig[] = [{ path: '/' }]
) {
  return new vr({ routes, mode: 'history' });
}

/**
 * Creates a form error message for testing.
 */
export function createFormAlertMessage(
  field: string,
  message: string
): FormAlertMessage {
  return { field, message };
}

/**
 * Creates a local vue instance with options to use any vue plugins.
 */
export function createLocalVueInstance(options: LocalVueSetupOptions = {}) {
  const { useVueCookies, useVuex, useVuetify, useVueRouter } = options;
  const localVue = createLocalVue();

  if (useVueCookies) {
    Vue.use(VueCookies);
  }
  if (useVuex) {
    localVue.use(Vuex);
  }
  if (useVuetify) {
    Vue.use(Vuetify);
  }
  if (useVueRouter) {
    localVue.use(VueRouter);
  }

  return localVue;
}
