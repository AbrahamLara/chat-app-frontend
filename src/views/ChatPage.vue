<template>
  <div class="chat-page d-flex fill-height">
    <v-row no-gutters>
      <v-col class="pa-0 fill-height col--hidden" sm="3">
        <div class="chat-page__sidebar fill-height">
          <v-card
            class="chat-page__sidebar-container fill-height d-flex flex-column"
            tile
            flat
          >
            <v-card-title class="chat-page__search-container pa-3">
              <v-text-field
                label="Search"
                prepend-inner-icon="mdi-magnify"
                outlined
                dense
                hide-details
                @keyup="handleUserSearch"
                autocomplete="off"
              ></v-text-field>
            </v-card-title>
            <!--suppress Stylelint -->
            <div style="flex: 1; overflow: hidden;">
              <user-list
                name="Results"
                :loading="isListLoading"
                :error-message="searchErrorMessage"
              >
                <div v-if="!users.length">
                  <div
                    class="text--secondary text-center mt-3"
                    v-text="'No results to display'"
                  ></div>
                </div>
                <template v-for="(user, index) in users">
                  <user-list-item
                    :key="index"
                    :user="user"
                    @click="handleUserItemClick"
                  ></user-list-item>
                  <v-divider :key="`divider-${index}`"></v-divider>
                </template>
              </user-list>
            </div>
          </v-card>
        </div>
      </v-col>
      <v-col class="pa-0 fill-height" sm="9">
        <div class="ChatPage__empty-container secondary">
          <div
            class="ChatPage__empty-container-message grey--text text--lighten-2 text-h4"
          >
            <div>Messages</div>
            <div>container</div>
          </div>
        </div>
      </v-col>
    </v-row>
  </div>
</template>

<script lang="ts">
import { Component } from 'vue-property-decorator';
import UserListItem from '@/components/UserListItem.vue';
import { UserItem } from '@/utils/misc-utils';
import UserList from '@/components/UserList.vue';
import tempStyles from '@/styles/ChatPageTempStyles.temp.sass';
import { isInvalidTokenMessage } from '@/utils/auth-utils';
import VueAuthHelper from '@/component-utils/VueAuthHelper';
import { Action } from '@/utils/decorators';
import { SET_ERRORS } from '@/store/constants/alerts-constants';
import { ALERTS_NAMESPACE } from '@/store/modules';
import { searchUser } from '@/api/search-api';

@Component({
  name: 'ChatPage',
  components: {
    'user-list': UserList,
    'user-list-item': UserListItem,
  },
})
export default class ChatPage extends VueAuthHelper {
  @Action(SET_ERRORS, ALERTS_NAMESPACE) setErrors!: Function;
  users: UserItem[] = [];

  /**
   * Indicates if the user list should display a spinner to convey a search being performed.
   */
  isListLoading = false;

  /**
   * The setTimeout id that will perform a search. The id is used to clear the setTimeout on every keyup event until
   * to prevent performing a search until the user stops typing.
   */
  timeoutID: ReturnType<typeof setTimeout> | null = null;

  /**
   * The error message to display when an error occurs performing a search.
   */
  searchErrorMessage = '';

  /**
   * Apply temporary styles that modify the application for this specific page.
   */
  beforeMount() {
    tempStyles.use();
  }

  /**
   * Remove temporary styles so that they don't remain set since they are only meant to be applied for this page.
   */
  beforeDestroy() {
    tempStyles.unuse();
  }

  /**
   * Handles a user item being clicked.
   */
  handleUserItemClick(user: UserItem) {
    console.log(user.name);
  }

  /**
   * This function handles user search and allows the user to make a user search by simply typing out a name and not
   * having to hit "Enter". This function utilizes the debouncing technique to prevent multiple api calls on every
   * keystroke in the search input.
   */
  handleUserSearch(event: any) {
    const search = event.target.value;

    if (!this.isListLoading) {
      this.isListLoading = true;
    }

    // Clear the error message displayed.
    if (this.searchErrorMessage) {
      this.searchErrorMessage = '';
    }

    // Clear the timeout function with the given id.
    if (this.timeoutID) {
      clearTimeout(this.timeoutID);
    }

    // If the search input is a non-empty string, perform a user search. Otherwise, clear the search results if there
    // are any left.
    if (search) {
      this.timeoutID = setTimeout(() => {
        this.userSearch(search);
      }, 600);
      return;
    } else if (this.users.length) {
      this.users = [];
    }

    // Display the loading spinner in the user list while a search is being performed.
    this.isListLoading = false;
  }

  /**
   * Handles making the search api call for users using the provided value.
   */
  async userSearch(value: string) {
    try {
      const response = await searchUser(value);
      const data = await response.json();

      if (response.ok) {
        this.users = data.results;
      } else if (isInvalidTokenMessage(data)) {
        this.unAuthenticateUser();
      } else {
        this.users = [];
      }
    } catch (event) {
      this.searchErrorMessage = 'An error occurred performing a search.';
    }

    // Set the loading flag to false to remove the spinner.
    this.isListLoading = false;
  }
}
</script>

<style lang="sass">
@use '../styles/vuetify-vars' as v

.ChatPage__empty-container
  display: flex
  justify-content: center
  align-content: center
  height: 100%
  text-align: center

  .ChatPage__empty-container-message
    height: fit-content
    margin-top: auto
    margin-bottom: auto

//noinspection Stylelint
@media screen and (max-width: map-get(v.$grid-breakpoints, 'sm'))
  .chat-page
    .col.col--hidden
      display: none
</style>
