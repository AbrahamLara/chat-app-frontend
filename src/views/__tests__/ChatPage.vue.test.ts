import {
  createLocalVueInstance,
  createRouter,
  createStore,
} from '@/utils/test-utils';
import { mount, Wrapper } from '@vue/test-utils';
import Vuetify from 'vuetify';
import ChatPage from '@/views/ChatPage.vue';
import Vuex, { Store } from 'vuex';
import { RootState } from '@/store/store-states';
import VueRouter from 'vue-router';
import UserList from '@/components/UserList.vue';
import { UserItem } from '@/utils/misc-utils';

const localVue = createLocalVueInstance({
  useVuetify: true,
  useVueRouter: true,
  useVueCookies: true,
  useVuex: true,
});

const MOCK_USER_SEARCH_RESULTS: UserItem[] = [
  {
    name: 'User 1',
    timeStamp: 1636416075544,
  },
  {
    name: 'User 2',
    timeStamp: 1636416075544,
  },
  {
    name: 'User 3',
    timeStamp: 1636416075544,
  },
];

describe('ChatPage.vue', () => {
  let wrapper: Wrapper<any>;
  let store: Store<RootState>;
  let router: VueRouter;

  beforeEach(() => {
    const vuetify = new Vuetify();
    store = createStore(Vuex);
    router = createRouter(VueRouter);

    wrapper = mount(ChatPage, {
      localVue,
      vuetify,
      store,
      router,
    });
  });

  it('renders the user search results', async () => {
    // There should only be a message conveying an empty list.
    expect(wrapper.findComponent(UserList).html()).toMatchSnapshot();

    // Mock the page having user search results.
    wrapper.vm.users = MOCK_USER_SEARCH_RESULTS;

    await wrapper.vm.$nextTick();

    // There should now be more 3 results in the list.
    expect(wrapper.findComponent(UserList).html()).toMatchSnapshot();
  });
});
