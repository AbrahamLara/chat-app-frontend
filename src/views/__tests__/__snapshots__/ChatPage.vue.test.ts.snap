// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatPage.vue renders the user search results 1`] = `
<div role="list" class="v-list user-list pa-0 d-flex flex-column fill-height v-sheet theme--light">
  <div tabindex="-1" role="listitem" class="v-list-item v-list-item--dense theme--light">
    <div class="v-list-item__title">Results</div>
  </div>
  <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
  <div class="user-list__items-container overflow-hidden flex">
    <div class="user-list__items-wrapper fill-height overflow-auto">
      <div>
        <div class="text--secondary text-center mt-3">No results to display</div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ChatPage.vue renders the user search results 2`] = `
<div role="list" class="v-list user-list pa-0 d-flex flex-column fill-height v-sheet theme--light">
  <div tabindex="-1" role="listitem" class="v-list-item v-list-item--dense theme--light">
    <div class="v-list-item__title">Results</div>
  </div>
  <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
  <div class="user-list__items-container overflow-hidden flex">
    <div class="user-list__items-wrapper fill-height overflow-auto">
      <!---->
      <div tabindex="-1" role="listitem" class="user-list-item v-list-item theme--light">
        <div class="v-avatar v-list-item__avatar" style="height: 50px; min-width: 50px; width: 50px; background-color: rgb(170, 170, 170); border-color: #aaa;"><i aria-hidden="true" class="v-icon notranslate mdi mdi-account theme--light"></i></div>
        <div class="v-list-item__content">
          <div class="d-flex justify-space-between">
            <div class="v-list-item__title">User 1</div> <span><div class="v-list-item__subtitle">Nov 8</div></span>
          </div>
          <!---->
        </div>
      </div>
      <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
      <div tabindex="-1" role="listitem" class="user-list-item v-list-item theme--light">
        <div class="v-avatar v-list-item__avatar" style="height: 50px; min-width: 50px; width: 50px; background-color: rgb(170, 170, 170); border-color: #aaa;"><i aria-hidden="true" class="v-icon notranslate mdi mdi-account theme--light"></i></div>
        <div class="v-list-item__content">
          <div class="d-flex justify-space-between">
            <div class="v-list-item__title">User 2</div> <span><div class="v-list-item__subtitle">Nov 8</div></span>
          </div>
          <!---->
        </div>
      </div>
      <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
      <div tabindex="-1" role="listitem" class="user-list-item v-list-item theme--light">
        <div class="v-avatar v-list-item__avatar" style="height: 50px; min-width: 50px; width: 50px; background-color: rgb(170, 170, 170); border-color: #aaa;"><i aria-hidden="true" class="v-icon notranslate mdi mdi-account theme--light"></i></div>
        <div class="v-list-item__content">
          <div class="d-flex justify-space-between">
            <div class="v-list-item__title">User 3</div> <span><div class="v-list-item__subtitle">Nov 8</div></span>
          </div>
          <!---->
        </div>
      </div>
      <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
    </div>
  </div>
</div>
`;
